nat-lab/tests/test_client_basic_stun.py

```diff
@@ -24,3 +30,70 @@ async def test_client_basic_stun(connection_tag: ConnectionTag, public_ip: str)
     async with new_connection_by_tag(connection_tag) as connection:
         ip = await testing.wait_long(stun.get(connection, config.STUN_SERVER))
         assert ip == public_ip, f"wrong public ip for the client {ip}"
+
+@pytest.mark.asyncio
+# Regression test for LLT-4306
+async def test_infinite_loop_in_stun_when_nonet() -> None:
+    async with AsyncExitStack() as exit_stack:
+        api = API()
+
+        (alpha, beta) = api.default_config_two_nodes()
+        alpha_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_1)
+        )
+        beta_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_2)
+        )
+
+        alpha_client = await exit_stack.enter_async_context(
+            Client(
+                alpha_connection,
+                alpha,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(
+                api.get_meshmap(alpha.id),
+            )
+        )
+        beta_client = await exit_stack.enter_async_context(
+            Client(
+                beta_connection,
+                beta,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(api.get_meshmap(beta.id))
+        )
+
+        await testing.wait_defined(
+            asyncio.gather(
+                alpha_client.wait_for_state_peer(
+                    beta.public_key, [State.Connected], [PathType.Direct]
+                ),
+                beta_client.wait_for_state_peer(
+                    alpha.public_key, [State.Connected], [PathType.Direct]
+                ),
+            ),
+            60,
+        )
+
+        for server in config.DERP_SERVERS:
+            await exit_stack.enter_async_context(
+                alpha_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
+            await exit_stack.enter_async_context(
+                beta_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
+
+        # 3478 and 3479 are STUN ports in natlab containers
+        tcpdump = await exit_stack.enter_async_context(
+            alpha_connection.create_process(
+                ["tcpdump", "-i", "any", "(", "port", "3478", "or", "3479", ")"]
+            ).run()
+        )
+        await asyncio.sleep(5)
+        
+        stun_requests = tcpdump.get_stdout().splitlines()
+        # There seems to be some delay when getting stdout from a process 
+        # Without this delay, `stun_requests` is empty even if tcpdump reports traffic
+        await asyncio.sleep(0.5)
```

Comment:
how sure are you, that this will not introduce flakiness? 😢

Reply:
Not very sure at all, but I first noticed that if I added a simple `print(...)` there the test worked and without it, it didn't, so it should not be flaky (famous last words)

Reply:
surely it will be fine